### PR TITLE
SubscriptionBond (inc. Websocket, more accurate notifications) 

### PR DESCRIPTION
**Pull request by kaikun213**
In progress. (still dependent on new parity-js pubsub feature)

### CODE REVIEW

**Comment by kaikun213** [comment]
For Bonds which take an input and are not likely that the result will change frequently we maybe should stay with the api calls. E.g. bonds.blockByNumber will likely change the given number often which would lead to a new unsubscribe/subscribe its much less likely that the returned block is another one (only reorgs). However a subscriptionBond insteadof api call will always send one more message (request-response-response + unsubscribe request-response) instead of simple request-response.

**Comment by gavofyork** [comment]
Best, I think, to keep everything consistent and correct for now with `SubscriptionBond` use throughout., 

If speed becomes an issue we can revisit later.

File: src/index.js [hunk 1]
```diff
@@ -65,22 +90,22 @@ function createBonds(options) {
...
 	}
 
-	// TODO: Use more generic means to check on number, ideally push notification.
 	class SubscriptionBond extends Bond {
-		constructor(rpc) {
+		constructor(rpc, options = []) {
 			super();
-			this.rpc = rpc;
+			this.rpc = rpc();
+			this.options = [(_,n) => this.trigger(n), ...options];
```
Comment (gavofyork) [inline src/index.js]: missing space after `,`

File: src/index.js [hunk 2]
```diff
@@ -65,22 +90,22 @@ function createBonds(options) {
...
+			this.subscription = this.rpc.apply(api().pubsub, this.options);
 		}
 		finalise () {
-			api().unsubscribe(this.subscription);
+			this.subscription.then(id => api().pubsub.unsubscribe([id]));
 		}
 		map (f) {
-	        return new TransformBond(f, [this]);
-	    }
+					return new TransformBond(f, [this]);
```
Comment (gavofyork) [inline src/index.js]: incorrect tabbing - `}` should be indented as much as the line that has the matching `{` on.

File: src/index.js [hunk 3]
```diff
@@ -188,19 +213,22 @@ function createBonds(options) {
...
-	bonds.blockNumber = new TransformBond(() => api().eth.blockNumber().then(_=>+_), [], [bonds.time]);
-//	bonds.blockNumber = new TransformBond(_=>+_, [new SubscriptionBond('eth_blockNumber')]);
-//	bonds.accounts = new SubscriptionBond('eth_accounts').subscriptable();
-//	bonds.accountsInfo = new SubscriptionBond('parity_accountsInfo').subscriptable();
-//	bonds.defaultAccount = new SubscriptionBond('parity_defaultAccount').subscriptable();
-//	bonds.allAccountsInfo = new SubscriptionBond('parity_allAccountsInfo');
-//	bonds.requestsToConfirm = new SubscriptionBond('signer_requestsToConfirm');
+	bonds.height = new TransformBond(_=>+_, [new SubscriptionBond(() => api().pubsub.blockNumber)]).subscriptable();
+	bonds.accounts = new SubscriptionBond(() => api().pubsub.accounts).subscriptable();
+	bonds.allAccountsInfo = bonds.accounts;	// double?
```
Comment (gavofyork) [inline src/index.js]: incorrect - `allAccountsInfo` is a completely different API.

File: src/index.js [hunk 4]
```diff
@@ -188,19 +213,22 @@ function createBonds(options) {
...
+	bonds.pendingTransactions = new SubscriptionBond(() => api().pubsub.pendingTransactions).subscriptable();
+	bonds.unsignedTransactionCount = new SubscriptionBond(() => api().pubsub.unsignedTransactionsCount).subscriptable();
+	//bonds.allAccountsInfo = new SubscriptionBond('parity_allAccountsInfo');
+	//bonds.requestsToConfirm = new SubscriptionBond('signer_requestsToConfirm');
 
 	Function.__proto__.bond = function(...args) { return new TransformBond(this, args); };
 	Function.__proto__.unlatchedBond = function(...args) { return new TransformBond(this, args, [], false, undefined); };
     Function.__proto__.timeBond = function(...args) { return new TransformBond(this, args, [bonds.time]); };
-    Function.__proto__.blockBond = function(...args) { return new TransformBond(this, args, [bonds.blockNumber]); };
+    Function.__proto__.blockBond = function(...args) { return new TransformBond(this, args, [bonds.height]); };
```
Comment (gavofyork) [inline src/index.js]: tabs not spaces (also the original line above needs correcting)

**Review by gavofyork — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 2 above)
Comment (gavofyork) [reply src/index.js]: also, i had replaced this with the implementation from above. that should be used now:

```
map (f, outResolveDepth = 0, resolveDepth = 1) {
 	        return new TransformBond(f, [this], [], outResolveDepth, resolveDepth);
}
```

File: src/index.js [hunk 5]
```diff
@@ -65,22 +90,22 @@ function createBonds(options) {
...
 		finalise () {
-			api().unsubscribe(this.subscription);
+			this.subscription.then(id => api().pubsub.unsubscribe([id]));
 		}
 		map (f) {
-	        return new TransformBond(f, [this]);
-	    }
+					return new TransformBond(f, [this]);
+			}
 		sub (name) {
```
Comment (gavofyork) [inline src/index.js]:
```
sub (name, outResolveDepth = 0, resolveDepth = 1) {
	return new TransformBond((r, n) => r[n], [this, name], [], outResolveDepth, resolveDepth);
}
```

**Comment by kaikun213** [comment]
@gavofyork parity-pubsub is merged in master. Could soon be merged?

**Comment by gavofyork** [comment]
just use #5.

**Pull request closed without merging**